src/cliffordsynthesis/encoding/SingleGateEncoder.cpp

```diff
@@ -36,16 +36,52 @@ void SingleGateEncoder::assertGateConstraints() {
     TRACE() << "Asserting gate constraints at time " << t;
     assertSingleQubitGateConstraints(t);
     assertTwoQubitGateConstraints(t);
+    assertNoGateNoChangeConstraint(t);
+  }
+}
+
+void SingleGateEncoder::assertNoGateNoChangeConstraint(std::size_t pos) {
+  for (std::size_t q = 0U; q < N; ++q) {
+    const auto noChange = createNoChangeOnQubit(pos, q);
+    const auto noGate   = createNoGateOnQubit(pos, q);
+    lb->assertFormula(LogicTerm::implies(noGate, noChange));
   }
 }
 
 void SingleGateEncoder::assertSingleQubitGateConstraints(std::size_t pos) {
   const auto& singleQubitGates = vars.gS[pos];
   for (std::size_t q = 0U; q < N; ++q) {
+    DEBUG() << "Asserting gates on " << q;
+
+    // Gates that leave qubit constant
+    lb->assertFormula(LogicTerm::implies(
+        singleQubitGates[gateToIndex(qc::OpType::None)][q] ||
+            singleQubitGates[gateToIndex(qc::OpType::X)][q] ||
+            singleQubitGates[gateToIndex(qc::OpType::Y)][q] ||
+            singleQubitGates[gateToIndex(qc::OpType::Z)][q],
+        (tvars->x[pos + 1][q] == tvars->x[pos][q]) &&
+            (tvars->z[pos + 1][q] == tvars->z[pos][q])));
+
+    // Hadamard changes both
+    lb->assertFormula(
+        LogicTerm::implies(singleQubitGates[gateToIndex((qc::OpType::H))][q],
+                           ((tvars->x[pos + 1][q] == tvars->z[pos][q]) &&
+                            (tvars->z[pos + 1][q] == tvars->x[pos][q]))));
+
+    // S and Sdag
+    lb->assertFormula(LogicTerm::implies(
+        (singleQubitGates[gateToIndex((qc::OpType::S))][q] ||
+         singleQubitGates[gateToIndex((qc::OpType::Sdag))][q]),
+        (tvars->z[pos + 1][q] == (tvars->z[pos][q] ^ tvars->x[pos][q]) &&
+         tvars->x[pos + 1][q] == tvars->x[pos][q])));
```

Comment:
Okay, I was initially against this but I managed to find a somewhat elegant way of generalizing this that has 0 redundancy (elegant if you like `std::function` anyway). If the `singleQubit#Change` functions in `TableauEncoder` change, than the grouping that is done in the `SingleGateEncoder` should be adapted accordingly.

The question is whether this functions are even necessary in the first place exactly because we want to have the mapping exactly the other way around. Once I also adjust the `MultiGateEncoder` I think the functions will be no longer called from anywhere.

Reply:
Looking forward to what you have came up with.
If some of the existing code can be removed/replaced, even better..